### ISSUE TITLE
Can't use IMAP/POP3 Tor onion service

### ISSUE THREAD

**Issue by yegortimoshenko** [mailpile/Mailpile#2035]
When I try to connect to POP3 or IMAP onion service, I'm getting the following error (1.0.0rc2):

```
Traceback (most recent call last):
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mail_source/__init__.py\", line 1024, in run
    self.sync_mail()
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mail_source/__init__.py\", line 293, in sync_mail
    elif (self._has_mailbox_changed(mbx_cfg, state) or
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mail_source/pop3.py\", line 114, in _has_mailbox_changed
    prefer_local=False)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/config/manager.py\", line 970, in open_mailbox
    mbox = msrc.open_mailbox(mbx_id, mfn.raw_fp) if msrc else None
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mail_source/pop3.py\", line 137, in open_mailbox
    throw=POP3_IOError)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mail_source/pop3.py\", line 46, in _open_pop3_mailbox
    debug=debug)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mailboxes/__init__.py\", line 68, in __init__
    parent.__init__(self, *args, **kwargs)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mailboxes/pop3.py\", line 65, in __init__
    self._connect()
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mailboxes/pop3.py\", line 81, in _connect
    self._pop3 = wrappable_POP3_SSL(self.host, self.port or 995)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/mailboxes/pop3.py\", line 32, in __init__
    self.sock = socket.create_connection((host, port), timeout)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/conn_brokers.py\", line 230, in create_brokered_conn
    address, self, self._need, self._reject, *a, **kw)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/conn_brokers.py\", line 682, in create_conn_with_caps
    address, context, need, reject, *args, **kwargs)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/conn_brokers.py\", line 302, in create_conn_with_caps
    *args, **kwargs)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/conn_brokers.py\", line 390, in _create_connection
    return self._conn(address, *args, **kwargs)
  File \"/nix/store/l11gvkih7lshavjnwh1ma0jlzmffic1x-mailpile-1.0.0rc2/lib/python2.7/site-packages/mailpile/conn_brokers.py\", line 461, in _conn
    sock.connect(address)
  File \"/nix/store/0lvikmnhhg6vlhcgidz0f6gjwyqw6jbs-python2.7-SocksiPy-branch-1.01/lib/python2.7/site-packages/socks.py\", line 392, in connect
    self.__negotiatesocks5(destpair[0],destpair[1])
  File \"/nix/store/0lvikmnhhg6vlhcgidz0f6gjwyqw6jbs-python2.7-SocksiPy-branch-1.01/lib/python2.7/site-packages/socks.py\", line 248, in __negotiatesocks5
    req = req + struct.pack(\">H\",destport).decode()
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 1: ordinal not in range(128)
```

To reproduce, create a fake new user account foo@riseup.net, disable "Detect settings", and fill connection information manually, pointing to 5gdvpfoh6kb2iqbizb37lzk2ddzrwa47m6rpdueg2m656fovmbhoptqd.onion as host name.

**Comment by BjarniRunar** [comment]
Interesting! I haven't tested things with .onion servers, so you are probably right it is broken. Thank you for reporting this!

**Comment by BjarniRunar** [comment]
I was unable to reproduce this. My Mailpile successfully connected to the Riseup IMAP server and POP3 servers using their .onion addresses.

I did have to use the short-form .onion addresses because my local Tor isn't new enough for the new style ones, but aside from that it's all working as far as I can tell. I tested both the socks and socksipychain libraries, both worked.

The only roadblock was I had to add a TOFU security exception for the IMAP server's TLS certificate, since obviously a *.riseup.net cert doesn't match a .onion name.

That error message you report is strange:

    File \"/nix/.../lib/python2.7/site-packages/socks.py\", line 248, in __negotiatesocks5
        req = req + struct.pack(\">H\",destport).decode()
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 1: ordinal not in range(128)

That just looks... weird. Why are they trying to decode a binary-packed 16 bit integer into ascii? I think your socks.py is broken.  I am quite tempted to close this as "not my fault", except now I want to know where that buggy socks.py comes from and how I can prevent Mailpile users from using it.
